src/Microsoft.Health.Dicom.SqlServer/Registration/ServiceCollectionRegistrationExtensions.cs

```diff
@@ -0,0 +1,145 @@
+            this IDicomFunctionsBuilder builder,
+            IConfiguration configurationRoot)
+        {
+            EnsureArg.IsNotNull(builder, nameof(builder));
+            IServiceCollection services = builder.Services;
+            SqlServerDataStoreConfiguration config = GetConfiguration(configurationRoot);
+            services.AddSingleton(Options.Options.Create(config));
+
+            SchemaInformation schemaInformation = new SchemaInformation(SchemaVersionConstants.Min, SchemaVersionConstants.Max);
+            services.AddSingleton(schemaInformation);
+
+            AddInitializedSqlServerBase(services, config);
+
+            services.AddScopedDefault<SqlInstanceStore>();
+
+            services.AddSqlExtendedQueryTagStores();
+
+            services.AddScopedDefault<SqlReindexStore>();
+            return builder;
+        }
+
+        public static IDicomServerBuilder AddSqlServer(
+          this IDicomServerBuilder dicomServerBuilder,
+          IConfiguration configurationRoot,
+          Action<SqlServerDataStoreConfiguration> configureAction = null)
+        {
+            EnsureArg.IsNotNull(dicomServerBuilder, nameof(dicomServerBuilder));
+            IServiceCollection services = dicomServerBuilder.Services;
+
+            services.AddSqlServerBase<SchemaVersion>(configurationRoot);
+            services.AddSqlServerApi();
+
+            var config = GetConfiguration(configurationRoot);
+
+            services.Add(provider =>
+                {
+                    configureAction?.Invoke(config);
+                    return config;
+                })
+                .Singleton()
+                .AsSelf();
+
+            services.Add(provider => new SchemaInformation(SchemaVersionConstants.Min, SchemaVersionConstants.Max))
+                .Singleton()
+                .AsSelf();
+
+            services.AddScopedDefault<
+                SqlIndexDataStoreV1,
+                SqlIndexDataStoreV2,
+                SqlIndexDataStoreV3,
+                SqlStoreFactory<ISqlIndexDataStore, IIndexDataStore>>();
+
+            // TODO: Ideally, the logger can be registered in the API layer since it's agnostic to the implementation.
+            // However, the current implementation of the decorate method requires the concrete type to be already registered,
+            // so we need to register here. Need to some more investigation to see how we might be able to do this.
+            services.Decorate<ISqlIndexDataStore, SqlLoggingIndexDataStore>();
+
+            services.AddScopedDefault<SqlQueryStore>();
+
+            services.AddScopedDefault<SqlInstanceStore>();
+
+            services.AddScopedDefault<SqlChangeFeedStore>();
+
+            services.AddSqlExtendedQueryTagStores();
+
+            return dicomServerBuilder;
+        }
+
+        private static void AddSqlExtendedQueryTagStores(this IServiceCollection services)
+        {
+            services.AddScopedDefault<
```

Comment:
Now I see why you want this method overload, but I think it's best just to separate this out into multiple calls. Actually, I think perhaps in the future we should refactor these "versioned" services so we can dynamically register all of the versions we find in the assembly.

But given we'll need more method overloads as we add new versions, I think it's best just to have separate method calls.